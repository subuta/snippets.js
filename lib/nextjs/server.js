import { build, snippets as s } from 'bld.js'
import _ from 'lodash'

export default () => {
  const imports = s.import([
    ['next', 'next'],
    ['koa', 'Koa'],
    ['koa-router', 'Router'],
    ['@koa/cors', 'cors'],
    ['koa-logger', 'logger'],
    ['koa-static', 'serve'],
    ['uuid/v4', 'uuid'],
    ['src/api/middlewares/zone.js', 'withZone'],
    ['src/api/routes', 'routes'],
    ['./config', null, [
      ['PUBLIC_DIR']
    ]],
  ])

  return build`
    ${imports}
    
    const port = parseInt(process.env.PORT, 10) || 3000
    const dev = process.env.NODE_ENV !== 'production'
    const app = next({
      dev,
      dir: 'src/views'
    })
    const handle = app.getRequestHandler()
    
    const startServer = () => {
      const server = new Koa()
      const router = new Router()
  
      // apply withZone to next.js routes.
      router.use(withZone)
  
      router.get('/p/:id', async ctx => {
        const actualPage = '/post'
        const queryParams = {id: ctx.params.id}
        await app.render(ctx.req, ctx.res, actualPage, queryParams)
        ctx.respond = false
      })
  
      router.get('*', async ctx => {
        await handle(ctx.req, ctx.res)
        ctx.respond = false
      })
  
      // log requests
      server.use(logger())
  
      // cors
      server.use(cors())
  
      // handle /api requests
      server.use(routes.routes())
  
      server.use(routes.allowedMethods())
  
      // otherwise PUBLIC_DIR
      server.use(serve(PUBLIC_DIR))
  
      server.use(async (ctx, next) => {
        ctx.res.statusCode = 200
        await next()
      })
  
      server.use(router.routes())
  
      if (!module.parent) {
        return server.listen(port, (err) => {
          if (err) throw err
          console.log(\`> Ready on http://localhost:\${port}\`)
        })
      }
    
      return server
    }
    
    if (!module.parent) {
      app.prepare()
        .then(startServer)
        .catch((ex) => {
          console.error(ex.stack)
          process.exit(1)
        })
      }
      
    export default startServer
  `
}
