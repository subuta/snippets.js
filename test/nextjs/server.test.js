/* global expect, describe, it */

import { build, format, snippets as s } from 'bld.js'

import server from 'lib/nextjs/server'

describe('nextjs/server.js', () => {
  it('should create server for next.js', () => {
    expect(format(server())).toMatchSnapshot()
  })
})
