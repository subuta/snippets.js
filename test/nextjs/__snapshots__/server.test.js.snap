// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nextjs/server.js should create server for next.js 1`] = `
"import next from 'next'
import Koa from 'koa'
import Router from 'koa-router'
import cors from '@koa/cors'
import logger from 'koa-logger'
import serve from 'koa-static'
import uuid from 'uuid/v4'
import withZone from 'src/api/middlewares/zone.js'
import routes from 'src/api/routes'
import {PUBLIC_DIR} from './config'

const port = parseInt(process.env.PORT, 10) || 3000
const dev = process.env.NODE_ENV !== 'production'
const app = next({
  dev,
  dir: 'src/views'
})
const handle = app.getRequestHandler()

const startServer = () => {
  const server = new Koa()
  const router = new Router()

  // apply withZone to next.js routes.
  router.use(withZone)

  router.get('/p/:id', async (ctx) => {
    const actualPage = '/post'
    const queryParams = {id: ctx.params.id}
    await app.render(ctx.req, ctx.res, actualPage, queryParams)
    ctx.respond = false
  })

  router.get('*', async (ctx) => {
    await handle(ctx.req, ctx.res)
    ctx.respond = false
  })

  // log requests
  server.use(logger())

  // cors
  server.use(cors())

  // handle /api requests
  server.use(routes.routes())

  server.use(routes.allowedMethods())

  // otherwise PUBLIC_DIR
  server.use(serve(PUBLIC_DIR))

  server.use(async (ctx, next) => {
    ctx.res.statusCode = 200
    await next()
  })

  server.use(router.routes())

  if (!module.parent) {
    return server.listen(port, (err) => {
      if (err) throw err
      console.log(\`> Ready on http://localhost:\${port}\`)
    })
  }

  return server
}

if (!module.parent) {
  app
    .prepare()
    .then(startServer)
    .catch((ex) => {
      console.error(ex.stack)
      process.exit(1)
    })
}

export default startServer
"
`;
